### ISSUE TITLE
Converting from flax back to pt?

### ISSUE THREAD

**Issue by orcinus** [cccntu/fine-tune-models#2]
Got the training working and it works really well for fine tuning out some common "glitches" and mis-reconstructions from photos and textures. However, i'd really like to convert the weights back into pytorch-usable format.

I've looked at the convert_diffusers_to_jax script from patil suraj's repo, but it doesn't look like reverse engineering it into an inverse function is feasible. Some changes it applies to the model appear to be irreversible without reconstructing the model key structure from scratch.

Diffusers now has its own pt->flax and flax->pt conversion, but i couldn't get its flax model to work with the script. There seem to be too many discrepancies in both the methods and the model keys and my knowledge is too limited to figure out how to adapt the fine tuning script to it. Diffuser's AutoEncoderKL also returns its output as yet another encapsulated proprietary class, complicating things even further.

So... bottom line - is there something i'm missing?
Is there some simple solution i've overlooked?

**Comment by orcinus** [comment]
Never mind, i figured out a viable way.

I've managed to shim the  diffuser's native FlaxAutoencoderKL to work with the script, and from there, it's a simple matter of juggling the pt and flax pipelines to get it converted (flax pipelines have a from_pt option, and pt pipelines have a from_flax option, so you can load pretrained from_pt into a flax autoencoder, train it, convert state to checkpoint, load pretrained from_flax into a pt autoencoder and dump into a pt checkpoint from there... not pretty, but it works).

**Comment by jbmaxwell** [comment]
@orcinus, I'm also in a situation where I have to fine-tune the stable diffusion VAE, but I'm also not running flax. Is there any chance you could share script/code or maybe a gist?... I'm sure a lot of people could benefit from that.

EDIT: Hmm... I think I misread your last post initially... so you managed to load Diffuser's FlaxAutoencoderKL. Makes sense... I'll poke around. But if you have anything you could share it would still be awesome! :)

**Comment by orcinus** [comment]
If you can hold for a bit longer, i need to tidy it up a bit first, then i'll put it up as a fork.
Also, yeah, once you get it running using FlaxAutoencoderKL, flax isn't an issue anymore, as you can go freely between flax and pytorch weights back and forth.

**Comment by omaralvarez** [comment]
I'm also interested in this, it would be of great help to see your code. By the way, thanks also for sharing your FlaxAutoencoderKL idea.

**Comment by orcinus** [comment]
Hit a bit of a snag...
I'm trying to get this to work for training at 512x512, but having difficulties.
There are also some other assorted issues to fix.

The problem with 512x512 is that lpips-j seems to be completely unsuitable for it (i'm guessing it was trained on 256x256) - it produces horrible artifacting on any textured or otherwise high-frequency detailed surface after only a couple of hundred iterations.

At the same time, L2, as used in the original repo, is also completely useless as it blurs the detail (as expected from L2). I'm guessing neither of those issues are a problem for training line art or other "non-complex" / "non-natural" distributions (manga, anime, illustrations), but when fine tuning off photos, the result is... ugly.

As to why insist on 512x512 - it's because from my test, it seems like ideally, fine tuning should be done in two stages, one for 512, one for 256, because there are larger scale problems to iron out of the VAE as well (local contrast issues, blue blotches, bad distribution in out of focus areas etc.).

**Comment by orcinus** [comment]
Actually, nvm, it appears that's a known issue with using LPIPS as the main driving loss:
![image](https://user-images.githubusercontent.com/1574258/234645113-a5a51f9a-79f9-4806-969b-7f7adac25d25.png)

And no, despite the claim, ti does not "train out".
It keeps getting worse.
The larger the dataset and the sharper the detail, the worse it gets.

I'll need to play around with schedules a bit to get it to NOT artifact up the reconstruction TOO much before the discriminator can kick in and take over. Essentially, plan is to probably switch from L2 to Huber, nerf both that and LPIPS so it just does minimal training for the first 1.5k-2k steps, then drive those two components off the cliff and let discriminator (mostly) take over.

Might be a super naive approach, i'm open to other suggestions.

**Comment by orcinus** [comment]
Now experimenting with combining FFT-based frequency domain loss(es) and spatial domain losses to see if that can help reduce the artifacting. I started playing with it randomly, then realized it's apparently already a thing in SRGAN circles...
